// @strapi/provider-email-nodemailer/dist/index.mjs
import nodemailer from 'nodemailer';

function getCapabilitiesFromOptions(opts) {
    const capabilities = {};
    const features = [];
    if (opts && typeof opts === 'object') {
        const options = opts;
        if (options.host || options.port) {
            capabilities.transport = {
                host: typeof options.host === 'string' ? options.host : undefined,
                port: typeof options.port === 'number' ? options.port : undefined,
                secure: typeof options.secure === 'boolean' ? options.secure : undefined,
                pool: typeof options.pool === 'boolean' ? options.pool : undefined,
                maxConnections: typeof options.maxConnections === 'number' ? options.maxConnections : undefined
            };
        }
        if (options.auth && typeof options.auth === 'object') {
            const auth = options.auth;
            const authType = typeof auth.type === 'string' ? auth.type : undefined;
            capabilities.auth = {
                type: authType || (auth.user ? 'login' : undefined),
                user: typeof auth.user === 'string' ? auth.user : undefined
            };
            if (authType === 'OAuth2') {
                features.push('oauth2');
            }
        }
        if (options.dkim) features.push('dkim');
        if (options.pool) features.push('pool');
        if (options.proxy) features.push('proxy');
        if (options.rateLimit) features.push('rateLimiting');
        if (options.requireTLS) features.push('requireTLS');
    }
    if (features.length > 0) {
        capabilities.features = features;
    }
    return capabilities;
}
var index = {
    init (providerOptions, settings) {
        const transporter = nodemailer.createTransport(providerOptions);
        return {
            send (options) {
                const message = {
                    from: options.from ?? settings.defaultFrom,
                    to: options.to,
                    cc: options.cc,
                    bcc: options.bcc,
                    replyTo: options.replyTo ?? settings.defaultReplyTo,
                    subject: options.subject,
                    text: options.text ?? options.html,
                    html: options.html ?? options.text
                };
                // Addressing
                if (options.sender) {
                    message.sender = options.sender;
                }
                // Content
                if (options.attachments) {
                    message.attachments = options.attachments;
                }
                if (options.alternatives) {
                    message.alternatives = options.alternatives;
                }
                if (options.watchHtml) {
                    message.watchHtml = options.watchHtml;
                }
                if (options.amp) {
                    message.amp = options.amp;
                }
                // Headers & metadata
                if (options.headers) {
                    message.headers = options.headers;
                }
                if (options.priority) {
                    message.priority = options.priority;
                }
                if (options.messageId) {
                    message.messageId = options.messageId;
                }
                if (options.date) {
                    message.date = options.date;
                }
                if (options.xMailer !== undefined) {
                    message.xMailer = options.xMailer;
                }
                // Threading
                if (options.inReplyTo) {
                    message.inReplyTo = options.inReplyTo;
                }
                if (options.references) {
                    message.references = options.references;
                }
                // Encoding
                if (options.textEncoding) {
                    message.textEncoding = options.textEncoding;
                }
                if (options.encoding) {
                    message.encoding = options.encoding;
                }
                if (options.normalizeHeaderKey) {
                    message.normalizeHeaderKey = options.normalizeHeaderKey;
                }
                // Advanced features
                if (options.icalEvent) {
                    message.icalEvent = options.icalEvent;
                }
                if (options.list) {
                    message.list = options.list;
                }
                if (options.envelope) {
                    message.envelope = options.envelope;
                }
                if (options.dkim) {
                    message.dkim = options.dkim;
                }
                if (options.attachDataUrls) {
                    message.attachDataUrls = options.attachDataUrls;
                }
                // Security
                if (options.disableUrlAccess) {
                    message.disableUrlAccess = options.disableUrlAccess;
                }
                if (options.disableFileAccess) {
                    message.disableFileAccess = options.disableFileAccess;
                }
                // Raw MIME
                if (options.raw) {
                    message.raw = options.raw;
                }
                // DSN and per-message auth (supported by nodemailer, not fully typed)
                if (options.dsn) {
                    message.dsn = options.dsn;
                }
                if (options.auth) {
                    message.auth = {
                        user: options.auth.user,
                        refreshToken: options.auth.refreshToken,
                        accessToken: options.auth.accessToken,
                        ...options.auth.expires != null ? {
                            expires: options.auth.expires
                        } : {}
                    };
                }
                return transporter.sendMail(message);
            },
            async verify () {
                return transporter.verify();
            },
            isIdle () {
                return typeof transporter.isIdle === 'function' ? transporter.isIdle() : true;
            },
            close () {
                transporter.close();
            },
            getCapabilities () {
                return getCapabilitiesFromOptions(providerOptions);
            }
        };
    }
};

export { index as default };
